### PR TITLE
Fixed Rect results for a sequenced range

### LINKED ISSUE #7838: Rect utility returns wrong sizes in case of sequenced range
## 📝 Provide detailed reproduction steps (if any)

While working on #7705 I found out that there are cases when balloon panel being mispositioned, due to the fact that the [`Rect`](https://ckeditor.com/docs/ckeditor5/20.0.0/api/module_utils_dom_rect-Rect.html) utility [cares only for the first range's client rectangle](https://github.com/ckeditor/ckeditor5/blob/f31092d65d97e91b21cd77bd12099f0b80d2b09e/packages/ckeditor5-utils/src/dom/rect.js#L81) (while it can contain multiple).

Example repro case:

1.  Open https://ckeditor.com/docs/ckeditor5/19.0.0/features/link.html
2.  Focus the first editor.
3.  Make the following selection `The [capital] city`
4.  Bold it.
5.  Make a wider selection, e.g. `The [capital city of` [`Malta`](https://en.wikipedia.org/wiki/Malta) `is the top destination this summer.]`
6.  Use the link button to open link balloon.

### ✔️ Expected result

The balloon points to the middle of selection.

### ❌ Actual result

The balloon point to the beginning of the selection.

Reason is that it contains multiple boxes, the first of which (on chrome from my testing is often) 0px width, thus it points to the very beginning.

## 📃 Other details

*   Browser: Any
*   OS: Any

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by mlewand**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (utils): `Rect` utility returns wrong sizes in case of a sequenced range. Closes #7838. 

Feature (utils): Introduced the `Rect#getBoundingRect()` method that returns a `Rect` instance containing all other rectangles. Closes #7858.

---

### Additional information

Initially I was even tempted to simplify it and use `Range.getBoundingClientRect()` to do the computation for us, but then I saw a plethora of issues that this method not always give reliable results:

*   #4969
*   #4952

### CODE REVIEW

File: packages/ckeditor5-utils/src/dom/rect.js [hunk 1]
```diff
@@ -406,3 +406,25 @@ function isBody( elementOrRange ) {
...
+		combinedRect.left = Math.min( combinedRect.left, rangeRects[ i ].left );
+		combinedRect.bottom = Math.max( combinedRect.bottom, rangeRects[ i ].bottom );
+		combinedRect.top = Math.min( combinedRect.top, rangeRects[ i ].top );
+	}
+
+	combinedRect.width = combinedRect.right - combinedRect.left;
+	combinedRect.height = combinedRect.bottom - combinedRect.top;
+
+	copyRectProperties( rect, combinedRect );
+}
```
Comment (niegowski) [inline packages/ckeditor5-utils/src/dom/rect.js:430]: In the table package we have a helper function that is very similar (the same purpose but different approach). Maybe we could unify those to one helper function? 

https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-table/src/utils/ui/contextualballoon.js#L123-L144

Comment (mlewand) [reply packages/ckeditor5-utils/src/dom/rect.js]: Good idea, added #7858 that will also be closed by this PR.

I had a small problem with the naming, first wanted to make it a `Range#expand()` member, but the name didn't feel right. Ended up with static `Range.getBoundingRect()` which feels just right :ok\_hand:

Also I simplified mentioned contextualballoon logic a little bit as the way how `createBoundingRect` [internal logic was implemented in the `getBalloonCellPositionData` function](https://github.com/ckeditor/ckeditor5/blob/1738d0f2296cfc9699cb1d78d7c459f60f1465e0/packages/ckeditor5-table/src/utils/ui/contextualballoon.js#L91-L93) was a little confusing.

Comment (niegowski) [reply packages/ckeditor5-utils/src/dom/rect.js]: I guess you mean `Rect` not `Range` 😉

File: packages/ckeditor5-utils/src/dom/rect.js [hunk 2]
```diff
@@ -381,6 +382,40 @@ export default class Rect {
...
+		}
+
+		if ( rectangleCount == 0 ) {
+			return null;
+		}
+
+		boundingRectData.width = boundingRectData.right - boundingRectData.left;
+		boundingRectData.height = boundingRectData.bottom - boundingRectData.top;
+
+		return boundingRectData;
```
[the remark this answers is not in the document]
Comment (mlewand) [reply packages/ckeditor5-utils/src/dom/rect.js]: Fixed in f863b19

**Pull request merged**